random/_types.ts

```diff
@@ -0,0 +1,23 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+
+/**
+ * A function implementing the same contract as `Math.random`, i.e. returning a
+ * random number in the range [0, 1).
+ */
+export type Prng = typeof Math.random;
```

Comment:
Otherwise, `Prng`-typed functions won't allow for arguments, as `Math.random()` doesn't have any arguments.
```suggestion
export type Prng = (...args: number[]) => number;
```

Reply:
I'm not sure they _should_ allow for arguments, and I'm not even sure what those arguments would represent? The idea of passing it as an option is that it's a drop-in replacement for `Math.random`. If further customization is needed then it should be a different named type (which also makes me wonder whether `Prng` is the right name for this type after all).

Reply:
For psuedo-random number generators, arguments would represent parameters for distributions. E.g. we could one day have the following added to the package.
```ts
export function randomExponential(lambda: number): number {
  if (lambda <= 0) {
    throw new RangeError(
      `Cannot compute exponential: lambda must be greater than 0, got ${lambda}`,
    );
  }

  return -Math.log(1 - Math.random()) / lambda;
}
```
Imagine similar functions like `randomBinomial()`, etc. I guess we can wait until more of these are implemented to see if `...args: number[]` makes most sense and just use `...args: any[]` in the meantime.

Reply:
Those can just be function-returning-functions, same as proposed seeded random (or class methods, same as current implementation). Also `n > 0` doesn't fulfil the contract as it doesn't guarantee results in `[0, 1)`, so it wouldn't be suitable as a drop-in replacement for `Math.random` (`n >= 1` would work though).

```ts
type Prng = typeof Math.random;

function randomExponential(lambda: number): Prng {
  if (lambda < 1) {
    throw new RangeError(`lambda must be at least 1, got ${lambda}`);
  }

  return () => {
    return -Math.log(1 - Math.random()) / lambda;
  };
}
```

Reply:
2nd order functions make sense when the returned function relies on state. I.e. `randomSeeded()`. However, something like `randomExponential()` wouldn't rely on state. So there's no real need for it to be a 2nd order function.

Reply:
In that case I think `{ prng: () => randomExponential(lambda) }` works fine. I'm strongly against the idea of `Prng` being `(...args) => number` because I don't understand the point of a type that doesn't represent a clear contract. Even if it was `(arg?: number) => number` it's not clear, as that single `number` argument could mean different things to different prng functions. And how does the calling function decide what that argument should be? Adding an additional `argsToPassToPrng` parameter just complicates things more.

Reply:
In C++,

* A [uniform random bit generator](https://en.cppreference.com/w/cpp/named_req/UniformRandomBitGenerator) generates uniformly distributed numbers.
* A [distribution](https://en.cppreference.com/w/cpp/named_req/RandomNumberDistribution) samples a value according to an associated probability function using a given uniform random bit generator.

The two concepts are conflated into one `Prng` type here, which seems to be the source of confusion in this discussion?

Reply:
Hmm... of the two, I wonder if there's really a use case for supplying a distribution as an option? It might make sense for some functions but it's less generally applicable than supplying a uniformly-distributed prng (which is generally useful for allowing seeding, and I guess cryptographic security too but with the usual caveats around that).

Reply:
> PRNGs are purely mathematical. Hence, them only accepting 0 to N optional arguments of type number. Allowing this option provides lots of flexibility while not complicating implementations.

I think this will be clearer with an example.

`randomBetween` takes a `Prng` as an option, which it calls internally with zero arguments, expecting the result to be in `[0, 1)`. That's a simple contract and still provides plenty of flexibility. For example,  our current `seeded` function requires a seed (a numeric argument, albeit a `bigint`), but that's provided to the HOF, not the returned function, so the returned function can easily be provided as a `Prng` option, as seen in the tests and doc examples.

If you have another, user-supplied PRNG function that takes 3 numeric arguments, you just need to supply those arguments prior to passing as an option. For example, with `function customRandom(a: number, b: number, c: number): number`, you provide that as `{ prng: () => customRandom(1, -65, 99) }`. If those numbers vary each time, perhaps it'll be more like `{ prng: () => customRandom(nextA(), nextB('x', 'y'), nextC({ prop: true })) }`. The interface is just as versatile as it needs to be for any given function. `randomBetween` doesn't need to care about or understand any of those additional arguments.

If, on the other hand, the `Prng` type accepts one or more arguments, _even if those are only numerical arguments_, where will those arguments be supplied from when `randomBetween` calls it? And what does `randomBetween` assume them to mean? A seed value? A number of steps to advance the PRNG each time? And how does it decide how many arguments to pass? Should it read `prng.length` to find out? What if `prng` is variable-length? What if it uses `arguments` internally so its `length` can't be relied on? And does _every_ function that takes a `Prng` option need to re-implement those same details? What if different calling functions make different assumptions? It's not just that it's nonsensical as a contract, it's not even clear what it _might_ mean or how that could be implemented in code.

Restricting the argument types to `number` does nothing to solve the above problems, yet it's still too restrictive for many possible implementations (e.g. you may want to supply a Uint8Array of seed/state bytes, a string to generate a seed from, a file path like `/dev/urandom` to read synchronously from, etc)... But as noted above, these can easily be supplied as config prior to passing as an option, without changing the `Prng` type at all.

Reply:
Ah, ok. I understand now. Yes, we only want functions that use `options.prng` to call `options.prng()` and nothing more. Let's keep as is then. Thanks for clearing up.